### ISSUE TITLE
spdx-CycloneDX conversion issues: single "target" artifact in CycloneDX

### ISSUE THREAD

**Issue by dn-scribe** [bom-squad/protobom#26]
Topic identified by @puerco :

Issue:
An SPDX may have many top-level components (I'll call them here artifacts), while CycloneDX has one: ```metatdata->component```
How should we translate such an SPDX ?

Example:
I have a release of an app that is made of multiple services, and for any reason I do not package them together as one ```.msi``` file.
I have a release that requires additional components installed by the customer, I want for my own vulnerability management to handle the whole release together.

Solutions:
1. Create a phantom artifact as the target, and create a component for each artifact both in the ```metadata->component->components field and as an additional ```dependency``` table.
2. Create a phantom artifact as the target and link all spdx artifacts to this one as additionad ```dependencies``` tables.
3. Create multiple CycloneDX documents, one for each artifact.
4. Fail translation.

Discussion:
Option 3 was ruled out due to complexity, and our anticipation for a tool that will carve out a component to a separte SBOM.
Option 4 poses a limitation on our tool.

Suggested solution:
Option 1. Enable user to control the virtual component content (or at least the ```name, version, type, group and licensing``` fields. default to ```"name":"virtual-component", "version":spdxId, "type":"virtual-component"```

Please comment\suggest

**Comment by dn-scribe** [comment]
Any comments here or we default to the "virtual top component" approach;
By default we should give it some minimal content:
- name
- files analyzed should be false 
- package verification code - omitted

name will default to the spdx name (see #6 ) or set by the user through configuraion.

**Comment by jspeed-meyers** [comment]
@dn-scribe: PR welcome!

Also, thinking out loud: Would this functionality be a better fit for `sbom-convert` than `protobom` itself? I think this issue was created when there was still a primitive CLI in `protobom`.

**Comment by houdini91** [comment]
Hi, i opened a PR For this issue.
The PR is not yet ready to be merged but i would love your discussion and review as there are details there we should consider. 

https://github.com/manifestori/protobom/pull/1

* To avoid the conflict the PR is currently pointed to @manifestori PR 72 branch.
Once https://github.com/bom-squad/protobom/pull/72 Is merged i will change the base.
@puerco @dn-scribe @manifestori @jspeed-meyers

**Comment by houdini91** [comment]
New PR on sbom-convert allows user to select a specific root or use a virutal one.
I am moving the changes up to the sbom tool as they require a graph  change (@puerco)

https://github.com/bom-squad/sbom-convert/pull/14

I would highly value any conversation on this so we can get the ball rolling on this

**Comment by puerco** [comment]
OK, I was noodling on this issue last week and I think I have a solution that won't break the graph and still can be used for bidirectional I/O.

To summarize the issue: When we have an SBOM describing more than one piece of software:
```mermaid
flowchart TD
    Document --> Node1
Document --> Node2
Document --> Node3
```
... we cannot easily express it in CDX as it only has one top-level element in the metadata section. It has a shape more or less like this when ingested into protobom:

```mermaid
flowchart TD
    Document --> TopLevelNode
TopLevelNode --> Node1
TopLevelNode --> Node2

```

...however, [the top-level element in CDX is not really required](https://cyclonedx.org/docs/1.4/json/#metadata_component). So we simply skip it. We skip it and add all the top-level elements in the component tree:

```mermaid
flowchart TD
   
    style TopLevelNode fill:#ccc,stroke:#999,stroke-width:2px,color:#666,stroke-dasharray: 5 5
   Document ---> Node1
   Document -.-> TopLevelNode("Top Level Node (Empty)")
   Document ---> Node2

```

Doing it this way, when traversing the bom elements, you still get the components in the tree first and you can still serialize the components back to the protobom graph back and forth.

If we want to privilege one of the components and promote it to the top-level component in the metadata we can pass its id in a new option.

**Comment by houdini91** [comment]
So correct me of I understood wrong your suggesting having the root element empty and pushing the graph as is under the components section ..?

**Comment by puerco** [comment]
Sorry Mikey I lost track of this discussion with all the travel lately.

>  your suggesting having the root element empty and pushing all top level elements in under the components section ..?

Correct. 

> Will this not create a awkward cdx?

My first hunch is that it creates an awkward cdx, but then the spec does not require it, so it _could_ be fine 

> If the user choses one to be the top level does it not make sense to only have it's graph component in the sbom?

I think you are right, you cannot elevate just one because it then implies that the others are contained within it. Ugh. This is the multidocument solution that is not really a solution.

> Do you expect this to happen as part of the library at the cdx seralizer level or higher up (like the cli) ?

In the CDX serializer. 

> My only worry is that an empty root can be confusing as it probably the first thing you be looking for when opening the sbom, taking more time to find out what is the top level components using the graph may be a little more tedious.

Well this should be done by machines so no need to worry about tedious work :)

> I understand the issue with the virtual root changing in the sbom graph but it may be useful to understanding the sbom.

Thinking about it some more, I think we should have the default in the serializer be the no-root approach. This will ensure any SBOM written will preserve the graph. And then, we can have an option to opt for the phantom root. And we can make the phantom root the default in the translation CLI. This gives us the benefit of not modifying the graph by default when working programmatically and at the same time we provide a more usable SBOM to folks using the CLI. WDYT?

> Another thought I had is use the root.component.components field in cdx to push the roots that way leave the top level root empty

Yeah that crossed my mind too but you still need to define a name and a type, so it is not empty and not really "invisible" when you generate the graph.

**Comment by houdini91** [comment]
> Thinking about it some more, I think we should have the default in the serializer be the no-root approach. This will ensure any SBOM written will preserve the graph. And then, we can have an option to opt for the phantom root. And we can make the phantom root the default in the translation CLI.

Agreed i can create a pr for us to review, 
Q: How would you expect the CLI to select the phantom option? Should we have a custom option only for CDX serlizer or should we add a option to all serlizers?
